scrypto-unit/src/test_runner.rs

```diff
@@ -103,53 +104,81 @@ impl Compile {
-            wasm_instrumenter: WasmInstrumenter::default(),
+    pub fn with_custom_genesis(mut self, genesis: SystemTransaction) -> Self {
+        self.custom_genesis = Some(genesis);
+        self
+    }
+
+    pub fn build(self) -> TestRunner {
+        let mut runner = TestRunner {
+            scrypto_interpreter: ScryptoInterpreter {
+                wasm_metering_config: WasmMeteringConfig::V0,
+                wasm_engine: DefaultWasmEngine::default(),
+                wasm_instrumenter: WasmInstrumenter::default(),
+            },
+            substate_store: TypedInMemorySubstateStore::new(),
+            state_hash_support: Some(self.state_hashing)
+                .filter(|x| *x)
+                .map(|_| StateHashSupport::new()),
+            intent_hash_manager: TestIntentHashManager::new(),
+            next_private_key: 1, // 0 is invalid
+            next_transaction_nonce: 0,
+            trace: self.trace,
         };
-        let mut substate_store = TypedInMemorySubstateStore::new();
-        let transaction_receipt = execute_transaction(
-            &mut substate_store,
-            &scrypto_interpreter,
+        let genesis = self
+            .custom_genesis
+            .unwrap_or_else(|| create_genesis(BTreeMap::new(), 1u64, 1u64));
+        runner.execute_transaction_with_config(
+            genesis.get_executable(vec![AuthAddresses::system_role()]),
             &FeeReserveConfig::default(),
             &ExecutionConfig::default(),
-            &genesis.get_executable(vec![AuthAddresses::system_role()]),
         );
-        let commit_result = transaction_receipt.expect_commit();
-        commit_result.outcome.expect_success();
-        commit_result.state_updates.commit(&mut substate_store);
-        Self {
-            scrypto_interpreter,
-            staged_substate_store_manager: StagedSubstateStoreManager::new(substate_store),
-            intent_hash_manager: TestIntentHashManager::new(),
-            next_private_key: 1, // 0 is invalid
-            next_transaction_nonce: 0,
-            trace,
+        runner
+    }
+}
+
+pub struct TestRunner {
+    scrypto_interpreter: ScryptoInterpreter<DefaultWasmEngine>,
+    substate_store: TypedInMemorySubstateStore,
+    intent_hash_manager: TestIntentHashManager,
+    next_private_key: u64,
+    next_transaction_nonce: u64,
+    trace: bool,
+    state_hash_support: Option<StateHashSupport>,
+}
+
+impl TestRunner {
+    pub fn builder() -> TestRunnerBuilder {
+        TestRunnerBuilder {
+            custom_genesis: None,
+            trace: true,
+            state_hashing: false,
         }
     }
 
     pub fn substate_store(&self) -> &TypedInMemorySubstateStore {
-        &self.staged_substate_store_manager.root
+        &self.substate_store
     }
 
     pub fn substate_store_mut(&mut self) -> &mut TypedInMemorySubstateStore {
-        &mut self.staged_substate_store_manager.root
+        &mut self.substate_store
```

Comment:
**Note from author:**

I noticed that "staging" was used here in a pretty degenerate way, so I refactored it into an exact equivalent (to avoid complexity); please shout if I missed something!

Reply:
Yeah, this makes sense. If we were previously just using the `root` property on it directly, then it didn't give it any effective testing.